engines/lab/resource.cpp

```diff
@@ -0,0 +1,430 @@
+			fileNameStrFinal = "GAME/";
+	} else if (filename.hasPrefix("MUSIC:")) {
+		if (_vm->getPlatform() != Common::kPlatformAmiga)
+			fileNameStrFinal = "GAME/MUSIC/";
+		else
+			fileNameStrFinal = "MUSIC/";
+	}
+
+	if (filename.contains(':')) {
+		while (filename[0] != ':') {
+			filename.deleteChar(0);
+		}
+
+		filename.deleteChar(0);
+	}
+
+	fileNameStrFinal += filename;
+
+	return fileNameStrFinal;
+}
+
+Common::File *Resource::openDataFile(const char *fileName, uint32 fileHeader) {
+	Common::File *dataFile = new Common::File();
+	dataFile->open(translateFileName(fileName));
+	if (!dataFile->isOpen())
+		error("openDataFile: Couldn't open %s (%s)", translateFileName(fileName).c_str(), fileName);
+
+	if (fileHeader > 0) {
+		uint32 headerTag = dataFile->readUint32BE();
+		if (headerTag != fileHeader) {
+			dataFile->close();
+			error("openDataFile: Unexpected header in %s (%s) - expected: %d, got: %d", translateFileName(fileName).c_str(), fileName, fileHeader, headerTag);
+		}
+	}
+
+	return dataFile;
+}
+
+Common::String Resource::readString(Common::File *file) {
+	byte size = file->readByte();
+	if (!size)
+		return nullptr;
+
+	Common::String str;
+	char c;
+	for (int i = 0; i < size; i++) {
+		c = file->readByte();
+		// Decrypt char
+		c = (i < size - 1) ? c - 95 : '\0';
+		str += c;
+	}
+
+	return str;
+}
+
+int16 *Resource::readConditions(Common::File *file) {
+	int16 i = 0, cond;
+	int16 *list = new int16[25];
+	memset(list, 0, 25 * 2);
+
+	do {
+		cond = file->readUint16LE();
+		if (i < 25)
+			list[i++] = cond;
+	} while (cond);
+
+	return list;
+}
+
+RuleList *Resource::readRule(Common::File *file) {
+	char c;
+	RuleList *rules = new RuleList();
+
+	do {
+		c = file->readByte();
+
+		if (c == 1) {
+			Rule *rule = new Rule();
+			rule->_ruleType = file->readSint16LE();
+			rule->_param1 = file->readSint16LE();
+			rule->_param2 = file->readSint16LE();
+			rule->_condition = readConditions(file);
+			rule->_actionList = readAction(file);
+			rules->push_back(rule);
+		}
+	} while (c == 1);
+
+	return rules;
+}
+
+void Resource::freeRule(RuleList *ruleList) {
+	if (!ruleList)
+		return;
+
+	for (RuleList::iterator rule = ruleList->begin(); rule != ruleList->end(); ++rule) {
+		freeAction((*rule)->_actionList);
+		delete[](*rule)->_condition;
+		delete *rule;
+		*rule = nullptr;
+	}
+
+	delete ruleList;
+	ruleList = nullptr;
+}
+
+Action *Resource::readAction(Common::File *file) {
+	char c;
+	Action *action = nullptr;
+	Action *prev = nullptr;
+	Action *head = nullptr;
+
+	do {
+		c = file->readByte();
+
+		if (c == 1) {
+			action = new Action();
+			if (!head)
+				head = action;
+			if (prev)
+				prev->_nextAction = action;
+			action->_actionType = file->readSint16LE();
+			action->_param1 = file->readSint16LE();
+			action->_param2 = file->readSint16LE();
+			action->_param3 = file->readSint16LE();
+
+			if (action->_actionType == SHOWMESSAGES) {
+				char **messages = (char **)malloc(action->_param1 * 4);
```

Comment:
I think it would be much better to just get rid of the double function of the _data field of Action. Why not just give it both string and list-of-strings fields, and use the right one depending on the actionType? Then you get type safety and automatic memory management for free.